### PR TITLE
Remove configuration that allows non-hash test ID

### PR DESCRIPTION
**Pull request by matsduf**
This PR resolves issue #691. The use of configuration key "force_hash_id_use_in_API_starting_from_id" has been deprecated by #694.

This PR simplifies the database engine codes considerable and removes a feature that we see no real use of.

### CODE REVIEW

File: t/test_DB_backend.pl
Comment (mattias-p) [inline t/test_DB_backend.pl:108]: If you split this up into two separate calls to ok() it'll be easier to know exactly which condition failed.

**Review by mattias-p — COMMENTED** [review]
I just skimmed the code changes and they look pretty good so far.
Have you though about the compatibility consequences?

**Comment by matsduf** [comment]
> Have you though about the compatibility consequences?

Can you clarify what compatibility consequence that you are thinking of?

**Comment by mattias-p** [comment]
I'm thinking of how people with existing installations will be affected by this change.

Is the new version a drop-in replacement for the old one with respect to this change? I guess Backend won't complain about the unknown option in the config file since we're don't really have any validation, but does the behavior change in any user-observable way for people with this option set to different values? If there is a change in behavior we should probably defer this removal until we have proper validation of the config file.

**Comment by mattias-p** [comment]
Also, #715 mentions this option.

**Comment by matsduf** [comment]
> I'm thinking of how people with existing installations will be affected by this change.

Referring to an old test will only work with hash_id. Otherwise the test history is available as before.

**Comment by mattias-p** [comment]
So in case any old users use this format they'll need to extract a translation table mapping integer id to hash id when upgrading. Chances are we don't have any such users, right? At the very least it should be mentioned in the release notes.

And in effect test ids matching `'^[0-9]$|^[1-9][0-9]{1,8}$` are still technically valid but they never match anything? We should probably update API.md and Validation.pm to reject those values.

But the numeric ids in the databases are just single-column primary keys? Those should probably be kept as is.

**Comment by matsduf** [comment]
> So in case any old users use this format they'll need to extract a translation table mapping integer id to hash id when upgrading. Chances are we don't have any such users, right? At the very least it should be mentioned in the release notes.

I should be mentioned, and I find it unlikely that we have those users.

> And in effect test ids matching `'^[0-9]$|^[1-9][0-9]{1,8}$` are still technically valid but they never match anything? We should probably update API.md and Validation.pm to reject those values.

We could validate, but I do not think it is necessary.

> But the numeric ids in the databases are just single-column primary keys? Those should probably be kept as is.

Yes, the numeric id should be kept as the unique identifier and is used for sorting.

**Comment by mattias-p** [comment]
> We could validate, but I do not think it is necessary.

As a matter of hygiene we really should. The only thing that comes out of accepting test ids in `^[0-9]$|^[1-9][0-9]{1,8}$` is people spending time trying to understand and support it, for no good reason.

**Comment by matsduf** [comment]
> As a matter of hygiene we really should. The only thing that comes out of accepting test ids in `^[0-9]$|^[1-9][0-9]{1,8}$` is people spending time trying to understand and support it, for no good reason.

I am not against adding such a validation. It should then be done after this has been merged.

**Comment by mattias-p** [comment]
> It should then be done after this has been merged.

That's fine with me.

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm
Comment (mattias-p) [inline lib/Zonemaster/Backend/DB/PostgreSQL.pm:145]: Have you considered updating this to use `?` instead of string interpolation?

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm
Comment (mattias-p) [inline lib/Zonemaster/Backend/DB/PostgreSQL.pm:173]: This looks strange to me. What does the `'hash_id'` argument do?

File: lib/Zonemaster/Backend/DB/SQLite.pm
Comment (mattias-p) [inline lib/Zonemaster/Backend/DB/SQLite.pm:191]: Have you considered updating this to use ? instead of string interpolation?

**Review by mattias-p — COMMENTED** [review]
This clean up is such an improvement.

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm
Comment (matsduf) [reply lib/Zonemaster/Backend/DB/PostgreSQL.pm:145]: Not really because I wanted to limit the change to what is needed. Would that be a considerable improvement? If so, I could do that change.

File: lib/Zonemaster/Backend/DB/SQLite.pm
Comment (matsduf) [reply lib/Zonemaster/Backend/DB/SQLite.pm:191]: Not really because I wanted to limit the change to what is needed. Would that be a considerable improvement? If so, I could do that change.

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm
Comment (mattias-p) [reply lib/Zonemaster/Backend/DB/PostgreSQL.pm]: Yes, it would definitely be an improvement. It makes it much easier to make code reviews with regard to SQL injections.

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm
Comment (mattias-p) [reply lib/Zonemaster/Backend/DB/PostgreSQL.pm:173]: Right, thanks. I was looking at the wrong definition.

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm
Comment (matsduf) [reply lib/Zonemaster/Backend/DB/PostgreSQL.pm]: In the updated commits the change is already implemented.

File: lib/Zonemaster/Backend/DB/SQLite.pm
Comment (matsduf) [reply lib/Zonemaster/Backend/DB/SQLite.pm:191]: In the updated commits the requested change is already implemented.

File: lib/Zonemaster/Backend/DB/PostgreSQL.pm
Comment (matsduf) [reply lib/Zonemaster/Backend/DB/PostgreSQL.pm:173]: I also wondered what is was, and had to look it up, so that is why I had a link to the documentation available.

**Comment by matsduf** [comment]
@mattias-p and @pnax, please review.

**Review by mattias-p — DISMISSED** [review]

**Comment by pnax** [comment]
I haven't review this PR yet, but looking at the the description, I agree that I'll need to review #715 once this is merged.

**Comment by mattias-p** [comment]
> I haven't review this PR yet, but looking at the the description, I agree that I'll need to review #715 once this is merged.

Either that or we merge #715 first and resolve the conflict as part of this PR instead.

File: t/test_DB_backend.pl
Comment (pnax) [inline t/test_DB_backend.pl:108]: `lenght` -> `length`

**Review by pnax — COMMENTED** [review]
Beside the typo, I like how this PR lightens the code.

**Comment by pnax** [comment]
> Either that or we merge #715 first and resolve the conflict as part of this PR instead.

This looks like a good idea, so that this PR does the same for every database which might produce a better code history. What do you think @matsduf ?

File: t/test_DB_backend.pl
Comment (matsduf) [reply t/test_DB_backend.pl]: Fixed.

**Comment by matsduf** [comment]
@mattias-p and @pnax, please approve again, if you approve.

**Comment by matsduf** [comment]
> > I haven't review this PR yet, but looking at the the description, I agree that I'll need to review #715 once this is merged.
> 
> Either that or we merge #715 first and resolve the conflict as part of this PR instead.

Either @matsduf or @mattias-p has to fix a PR. I am open for either way.

**Comment by matsduf** [comment]
Or else I could "pick" the code in #715 that adds missing "overall" to SQLite and add that to this PR.

**Comment by pnax** [comment]
If it does not bother you, I think it would be nice to harmonize and then remove so that this removal is consistent across all database.

**Comment by pnax** [comment]
I merged #715.

**Comment by matsduf** [comment]
@mattias-p and @pnax, moved to updated develop branch. There were several conflicts in `SQLite.pm` that I had to fix manually. I found no issues when doing test with it, but please inspect that again with your eyes.

File: t/test_DB_backend.pl
Comment (pnax) [inline t/test_DB_backend.pl:105]: Parenthesis mismatch here.

File: t/test_DB_backend.pl
Comment (pnax) [inline t/test_DB_backend.pl:108]: same parenthesis problem

**Review by pnax — COMMENTED** [review]
Thank you for the fixes with SQLite.pm. Just a last small adjustment with some parentheses.

File: t/test_DB_backend.pl
Comment (matsduf) [reply t/test_DB_backend.pl]: I would say *mis*placed rather than *mis*match. Corrected. The code was legal, but what came after comma was meaningless.

File: t/test_DB_backend.pl
Comment (matsduf) [reply t/test_DB_backend.pl]: As above.

File: t/test_DB_backend.pl
Comment (pnax) [reply t/test_DB_backend.pl]: yes you are totally right, I couldn't find that word when commenting yesterday ;)

**Review by pnax — APPROVED** [review]

File: lib/Zonemaster/Backend/DB/SQLite.pm
Comment (mattias-p) [reply lib/Zonemaster/Backend/DB/SQLite.pm:192]: Did you accidentally revert this when resolving the conflicts?

**Review by mattias-p — APPROVED** [review]

File: lib/Zonemaster/Backend/DB/SQLite.pm
Comment (matsduf) [reply lib/Zonemaster/Backend/DB/SQLite.pm:192]: I probably never replaced "$minutes" with "?". That can be fixed later.

**Comment by mattias-p** [comment]
## Release testing for v2021.1

I verified this using the commands below on CentOS 7 and CentOS 8.

```
 # Under [ZONEMASTER] add the line "force_hash_id_use_in_API_starting_from_id = 10":
sudoedit /etc/zonemaster/backend_config.ini
sudo systemctl restart zm-rpcapi
sudo systemctl restart zm-rpcapi
journalctl -xe | grep force_hash_id_use_in_API_starting_from_id
# Verify that you got error messages about
# ZONEMASTER.force_hash_id_use_in_API_starting_from_id being unrecognize
# from both zm-rpcapi and zm-testagent
```

**Pull request merged**
